### DISCUSSION TITLE
go blocks in handlers?

### DISCUSSION THREAD

**Discussion by yayitswei** [day8/re-frame#707] [Q&A]
Is it ok to have a go block inside a handler?

My use case is a secretary handler that must wait for data from the server, then do some other stateful stuff. I'm planning to have something like `(go (<! loaded-from-server) (stateful-stuff!))` inside the secretary handler, since it's one of the first things that's run and must wait for the data. 

If this isn't idiomatic, please suggest a better method! Thanks for making re-frame.

**Comment by danielcompton** [comment]
A few thoughts
1. go blocks are asynchronous, so if you need to wait for server data before continuing then you'll need to block on the result before continuing
2. `dispatch-sync` lets you wait for the dispatch to complete before continuing
3. Generally you want to have your state all in the re-frame app-db, it should probably go there instead of living a handler on a go block. 
4. Short lived go blocks are fine in handlers, I'm not sure about long lived go blocks, mainly because I'd expect all state to go in the app-db and they wouldn't be required.
5. I'm pretty new to re-frame so this may not be right 'answer'.

**Comment by yayitswei** [comment]
Thanks for your answer, @danielcompton. In my case, the stateful stuff is manipulating an embedded YouTube player, so it can't be in the app-db. I believe taking from a port does block, so I'd receive the server data before doing any operations.

Will look into `dispatch-sync`.

**Comment by mike-thompson-day8** [comment]
I'm wondering if this helps:  https://github.com/Day8/re-frame/wiki/Talking-To-Servers  ?

Event handlers are:
- given the current value in `app-db`
- will mutate that current state in some way 
- are expected to return the new version of the state

And, in addition:
- might further `dispatch`  ([but never `dispatch-sync`](https://github.com/Day8/re-frame/wiki/FAQ#3-why-cant-i-call-dispatch-sync-in-an-event-handler))

So, yes, you could put a `go` block in an event handler:
-  but any async results would have to be handled via a further `dispatch`.
-  and, be aware that the event handler has to return a new value for app-db in-the-now (perhaps unchanged)

**Comment by yayitswei** [comment]
Thanks for the clarifications. I ended up going with a channel `initialized?` which gets closed once all the initial data is loaded and the YouTube player is ready. This way handlers that depend on the initial data can take from `initialized?` to block. As you suggested, those handlers would return an unchanged db and dispatch a new event inside the go block, after the take. Everything works as expected and the only downside is that I have a floating global `(def initialized? (chan))` in my `handlers.cljs`.

**Comment by mike-thompson-day8** [comment]
That wouldn't be my recommended way of doing it.  You want "state" in the one place. Another channel is really state in a second place.

Have you read this Wiki page:  https://github.com/Day8/re-frame/wiki/Bootstrap-An-Application

Notice the "Scales Up" section.

**Comment by yayitswei** [comment]
There's an example of a subscription that triggers on one or more values in app-db, but can I trigger a handler when a value changes? I haven't seen an idiomatic re-frame example yet which is why I'm using the go block. I can also pass on the original parameters to the blocking handler to the new handler, which I can't think of a good way to do without the go block. Possibly storing some temporary state (e.g. `item-to-display-on-load`) in app-db, but that seems a little awkward.

**Comment by mike-thompson-day8** [comment]
Hmm.  I'm not sure you are thinking about this the right way. 

You don't want to trigger a handler when a value changes.  You trigger event handlers when an event occurs.  

Ask yourself: what event causes the value to change?  Then work out how to `(dispatch [:some-id  X Y Z])` in response to that event occurring (this could be done in a go block).  

The  event handler for `some-id` will be responsible for updating values in `app-db`. Those changes then ripple to the views via FRP.   

I hope I'm helping here.
